pkg/aks/create.go

```diff
@@ -23,17 +24,162 @@ func CreateResourceGroup(ctx context.Context, groupsClient *resources.GroupsClie
 	return err
 }
 
-// CreateOrUpdateCluster creates a new managed Kubernetes cluster
-func CreateOrUpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+// CreateCluster creates a new managed Kubernetes cluster. In this case, there will be no existing upstream cluster.
+// We are provisioning a brand new one.
+func CreateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
 	spec *aksv1.AKSClusterConfigSpec, phase string) error {
 
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	_, err = clusterClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		spec.ClusterName,
+		managedCluster,
+	)
+
+	return err
+}
+
+// UpdateCluster updates an existing managed Kubernetes cluster. Before updating, it pulls any existing configuration
+// and then only updates managed fields.
+func UpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient,
+	spec *aksv1.AKSClusterConfigSpec, phase string) error {
+
+	// Create a new managed cluster from the AKS cluster config
+	managedCluster, err := CreateManagedCluster(ctx, cred, spec, phase)
+	if err != nil {
+		return err
+	}
+
+	// Pull the upstream cluster state
+	clusterUpdate, err := clusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
```

Comment:
The actual state of the cluster in the provide is already available in the calling function in the controller, so there is no need to retrieve it again https://github.com/rancher/aks-operator/blob/a32c2bf0c3254301d1257345d4db786fef49162e/controller/aks-cluster-config-handler.go#L702

A lot of the logic for determining what fields to update and what fields are unmanaged is already contained in that updateUpstreamClusterState method. Can you clarify what is different here, and why it is needed? Is this following the principle of using nil as a flag for what fields are managed and unmanaged? https://github.com/rancher/rancher/wiki/Kontainer-Engine-v2:-The-new-way-rancher-integrates-with-cloud-providers

Reply:
Good catch, but what if there's been an update in AKS since the upstreamSpec was last built?

Sure - that's true. For a lot of `updateUpstreamClusterState,` we just compare with the upstream state and then update upstream with whatever data is in the config spec. Thus, if a field in Rancher is nil and the upstream value is not the upstream value often gets overwritten. `updateCluster` in `create.go` addresses that by merging in data from upstream before the update is done. I could probably add that logic to `updateUpstreamClusterState` but that may make the operator less readable. Although it would leave just the create logic left in `create.go` which I actually might like better.. pending discussion